### PR TITLE
Issue 247  astra geometry slow

### PR DESCRIPTION
**Pull request by adler-j**
Minor changes according to #247 

* Add a cache to geometries where implementations can be stored. Running a 50^3 -> 50^3 problem went from 0.08s per iteration to 0.027s.
* Improve performance of `IntervalProd.__contains__`, reduce runtime by ~70%.

### CODE REVIEW

File: odl/set/domain.py [hunk 1]
```diff
@@ -154,9 +154,37 @@ def max(self):
...
+        >>> interv.element(0.5)
+        0.5
+        """
+        if inp is None:
+            return self.midpoint
+        elif inp in self:
+            return inp
+        else:
+            raise TypeError('inp {!r} not a valid element in {!r}.'
+                            ''.format(inp, self))
```
Comment (kohr-h) [inline odl/set/domain.py]: Reason for this change?

File: odl/set/domain.py [hunk 2]
```diff
@@ -218,23 +246,44 @@ def approx_contains(self, point, tol):
...
+        >>> interv = IntervalProd(0, 1)
+        >>> 0.5 in interv
+        True
+        >>> 2 in interv
+        False
+        >>> 'string' in interv
+        False
+        """
+        try:
+            # Ducktyped check
```
Comment (kohr-h) [inline odl/set/domain.py]: Duck-typed I think it is

File: odl/set/domain.py (hunk 1 above)
Comment (adler-j) [inline odl/set/domain.py]: Consistency, the definition of `element`says it should take an `inp`:

https://github.com/odlgroup/odl/blob/eec730cd0128a77a4315bf2303da755878bc346d/odl/set/sets.py#L126-L131

File: odl/tomo/backends/astra_setup.py [hunk 3]
```diff
@@ -396,6 +401,10 @@ def astra_projection_geometry(geometry):
         raise NotImplementedError('unkown ASTRA geometry type {}.'.format(
             geometry))
 
+    if 'astra' not in geometry.implementations:
+        # Shortcut, reuse already computed value.
```
Comment (kohr-h) [inline odl/tomo/backends/astra_setup.py]: Other way around

File: odl/set/domain.py (hunk 1 above)
Comment (kohr-h) [inline odl/set/domain.py]: OK

File: odl/set/domain.py [hunk 4]
```diff
@@ -218,23 +246,44 @@ def approx_contains(self, point, tol):
...
+        return self.dist(point, ord=np.inf) <= tol
 
     def __contains__(self, other):
-        """Return ``other in self``."""
-        return self.approx_contains(other, tol=0)
+        """Return ``other in self``.
+
+        Parameters
+        ----------
+        other
```
Comment (kohr-h) [inline odl/set/domain.py]: With or without colon at the end? Have you tested the doc?

Comment (adler-j) [inline odl/set/domain.py]: Nope, but this should work I think, will do a big test of doc before we merge conebeam

File: odl/tomo/geometry/geometry.py [hunk 5]
```diff
@@ -154,6 +155,18 @@ def ndim(self):
         return self._ndim
 
     @property
+    def implementations(self):
```
Comment (kohr-h) [inline odl/tomo/geometry/geometry.py]: Why "implementations"? What about "cache"?

Comment (adler-j) [inline odl/tomo/geometry/geometry.py]: I was thinking something complicated, so users dont go for it. How about `implementation_cache`?

Comment (kohr-h) [inline odl/tomo/geometry/geometry.py]: What about not making it public then?

Comment (adler-j) [inline odl/tomo/geometry/geometry.py]: Well its used by our other classes, private means only for this class.

Comment (kohr-h) [inline odl/tomo/geometry/geometry.py]: Alright, we should actually not use private attributes too much externally. Then s.th. like `impl_cache` would sound even more mysterious.

**Comment by kohr-h** [comment]
We have a failing test, actually for a while, in the conebeam branch:

```
Error Message

TypeError: unorderable types: DiscreteLpVector() > int()

Stacktrace

@skip_if_no_astra
    def test_astra_cpu_projector_fanflat():
        """ASTRA CPU forward and back projection for fanflat geometry."""
    
        # `DiscreteLp` space for volume data
        vol_shape = (4, 5)
        discr_vol_space = odl.uniform_discr([-4, -5], [4, 5], vol_shape,
                                            dtype='float32')
    
        # Create an element in the `DiscreteLp` space
        discr_data = odl.util.phantom.cuboid(discr_vol_space, 0.5, 1)
    
        # Angles
        angle_intvl = odl.Interval(0, 2 * np.pi)
        angle_grid = odl.uniform_sampling(angle_intvl, 8)
    
        # Detector
        dparams = odl.Interval(-6, 6)
        det_grid = odl.uniform_sampling(dparams, 6)
    
        # Distances for fanflat geometries
        src_rad = 100
        det_rad = 10
    
        # 2D geometry instances for parallel and fan beam with flat line detector
        geom = odl.tomo.FanFlatGeometry(angle_intvl, dparams, src_rad, det_rad,
                                        angle_grid, det_grid)
    
        # Projection space
        proj_space = odl.FunctionSpace(geom.params)
    
        # `DiscreteLp` projection space
        proj_shape = geom.grid.shape
        discr_proj_space = odl.uniform_discr_fromspace(proj_space, proj_shape,
                                                       dtype='float32')
    
        # forward
        discr_data = odl.util.phantom.cuboid(discr_vol_space, 0.5, 1)
        proj_data = odl.tomo.astra_cpu_forward_projector_call(discr_data, geom,
                                                              discr_proj_space)
        assert proj_data.shape == proj_shape
>       assert proj_data > 0
E       TypeError: unorderable types: DiscreteLpVector() > int()

test/tomo/backends/astra_cpu_test.py:124: TypeError
```

I wrote a comment in another place some time ago, but nobody fixed it.

File: odl/tomo/geometry/geometry.py (hunk 5 above)
Comment (adler-j) [inline odl/tomo/geometry/geometry.py]: Went for `implementation_cache` since its long and cumbersome :)

**Comment by adler-j** [comment]
Is that failing test really due to this branch?

**Comment by kohr-h** [comment]
No, it's from upstream. But as I wrote, there was no reaction.

**Comment by adler-j** [comment]
In general I think those tests are waiting for the correct scalings before we can do them. The `astra_scaling` branch contains fixes to this I think.

**Comment by kohr-h** [comment]
Yes, but this particular test fails because it's nonsense, regardless of scaling :-)

File: odl/tomo/geometry/geometry.py (hunk 5 above)
Comment (kohr-h) [inline odl/tomo/geometry/geometry.py]: Go for it!

**Comment by moosmann** [comment]
The test should be
```
assert proj_data.norm() > 0
```
and simply check that the projection data is positive. I already fixed this test in the `astra_scaling` branch but not in the cone-beam branch.

**Comment by kohr-h** [comment]
You're lucky that this dates back to the time when I was remodeling our Jenkins CI. Otherwise you would have gotten angry e-mail :-)

**Pull request merged**